WIP Box Rework

arakis opened the pull request:
### Purpose of this PR

As discussed in #1033, Box structs missing some functions to being more complete.

This is a first Draft. So I made changes first for `Box2`. All the other box structs will follow, incl. comments/documentation, and unit tests.

* Constructor and the current Interfaces are completely untouched.
* Added several Helpers and Aliases.

### Testing status

Not tested yet. I will write several unit tests once the rework is complete.

### Comments

We should discuss about the following Property: `Size`.
A classical `Rectangle` struct allows the resizing via Size-Setter, keeping the `Min` and increases `Max`. In the current work, the setter of `Width` and `Height` works exactly as expected from those structs.

But the current Size has it's origin in the Center, so setting it will modify `Min` and `Max`.

Because it's very common that you just want to set the size in a classic way, my thoughts:
* Introducing a new property, where the setter acts like a classic rectangle. Problem: Naming :-)
OR
* Rename the current Size to let's say SizeCentered / CenteredSize. 

Better suggestion? Let me hear!

NogginBops commented on src/OpenToolkit.Mathematics/Geometry/Box2.cs:
```diff
@@ -121,6 +121,160 @@ public Vector2 Center
+
+        public Vector2 Location => _min;
+
+        public bool IsEmpty => _min.X == 0 && _min.Y == 0 && _max.X == 0 && _max.Y == 0;
```
I think we should consider a different name for this, because I think IsEmpty can be confusing if you are doing stuff like collision-detection, and the name doesn't really say what it returns (a box with nothing in it could also be considered empty).

I think we could also add a `IsDegenerate` property to check if one of the sides have zero length.

arakis replied:
What about IsZero?

NogginBops replied:
I would be fine with that I guess. Another name could be `HasZeroSize` or something that says it's size is zero, `IsAreaZero` could also be an idea.

NogginBops commented on src/OpenToolkit.Mathematics/Geometry/Box2.cs:
```diff
@@ -121,6 +121,160 @@ public Vector2 Center
+
+        public static readonly Box2 Zero = new Box2(0, 0, 0, 0);
+
+        public static readonly Box2 Identity = new Box2(0, 0, 1, 1);
```
`Identity` is probably not the best name here. My vote is for `UnitBox`, `UnitSquare` or `UnitRect` as that conveys better what it is.

arakis replied:
Good Point. I will name it `UnitSquare`

jvbsl replied:
tbh I would expect a `UnitSquare` to be centered at the origin, but I can't think of a better name either.

NogginBops replied:
This is not what is expected in math terms, there is the square from 0,0 to 1,1.
https://en.wikipedia.org/wiki/Unit_square

jvbsl replied:
yeah then it is the perfect name. Was just expecting it that way because of the unit circle/spheres being centered at the origin.

NogginBops commented on src/OpenToolkit.Mathematics/Geometry/Box2.cs:
```diff
@@ -121,6 +121,160 @@ public Vector2 Center
+            return new Box2(minX, minY, maxX, maxY);
+        }
+
+        public void Intersect(Box2 other)
```
There should also exist a `Box2 Intersected(Box2 other)` overload to match the rest of the math APIs.

arakis replied:
Good Point. Will be added.

deccer replied:
but `Intersected` sounds like it would return `bool` not another `Box`. Name should be `Intersect` then.

NogginBops replied:
`Intersects` is the name I would expect on a function that returns a `bool`.

NogginBops commented on src/OpenToolkit.Mathematics/Geometry/Box2.cs:
```diff
@@ -121,6 +121,160 @@ public Vector2 Center
+                && _min.Y <= other._max.Y;
+        }
+
+        public static Box2 Union(Box2 a, Box2 b)
```
I would also like to see a `Box2 Inflate(Vector2 pos)` which extends the box to include the point if it's not already in the box.

I know `Union` is the correct name for this but personally I would have named this `Inflate` to match behaviors between the two functions (they both make the box bigger). But that's just my preference, both naming schemes work.

arakis replied:
We already have a `void Inflate(Vector2 point)` and `Box2 Inflated(Vector2 point)`, so that's not part of my patch.

Regarding `Union (Box2 a, Box2 b)`: I prefer `Union` but if somebody else thinks `Inflate` is the better choice, we can change it.

NogginBops replied:
I missed the fact that there existed an `Inflate(...)` already...
I'm fine with both names, I was just expressing my preference.

jvbsl replied:
Well a `Union` in my head would be more of a:
```
UnionBox2D
{
 List<Box2D> boxes;
 bool Intersect(Vector2 point) => boxes.Any(x => x.Intersect(point));
}
```
Therefore I think inflate still matches better.

NogginBops commented on src/OpenToolkit.Mathematics/Geometry/Box2.cs:
```diff
@@ -121,6 +121,160 @@ public Vector2 Center
+                && _min.Y < other._max.Y;
+        }
+
+        public bool TouchWith(Box2 other)
```
I haven't thought too much about the `<=` or `<` when doing `Contains` and similar.
I think we should establish a consistent representation (write a documentation on the type on how it's expected to work) and make sure all functions follow that way of thinking.
Basically; should the edge of the box be contained or not?

arakis replied:
`IntersectsWith` means, they need a real intersection, where `TouchWith` means, it's also ok if they are directly next to each other (they are aligned on the edge).

We already have a `Contains(Vector2 point, bool boundaryInclusive)` and i'm not happy with the extra bool. The main problem here is naming. I think Contains should/can include the border, but what name for, if it should be NOT on the border?

NogginBops replied:
I still think that an important part of this is good documentation so that you can figure this out regardless of naming. For this type of situation I would probably take a look at the documentation to figure out what the alternatives are. But yeah we want good names for these... (maybe there is a good math terminology for this)
And if we get good naming we can also remove the current `Contains(Vector2 point, bool boundaryInclusive)`.

NogginBops commented on src/OpenToolkit.Mathematics/Geometry/Box2.cs:
```diff
@@ -121,6 +121,160 @@ public Vector2 Center
 
+        // --
+
+        public float Width
```
Should there be a way to change the Width and Height at the same time like with `Size` but only move the max instead of resizing from the center?

frederikja163 commented:
Thanks for the great PR so far, Noggins review seems quite thorough to me, so if you go through with the changes he requested i think this is a addition to the boxes.

varon commented:
Nice work @arakis  - Thanks for the review @NogginBops .

I believe that master has fixes for CI. Can you try merge the latest master in, @arakis ?

Only thing I'd like to add would be a request for some extra unit tests to verify properties on these things - Figuring those out would be pretty fun I suspect.

arakis commented:
What I have changed so far:

* Renamed Size to CenteredSize (setter: min/max behavior) and added Size again (setter: pos,size behavior). Getter's behavior is the same.
* Renamed `IsEmpty` to `IsZero`.
* Renamed `Identity` to `UnitSquare`.
* Added `Box2 Intersected (Box2 other)`
* Added some XML comments, some are still missing.

The Question of the day is: Should change the constructor of from `Box2(min,max)` to `Box2(pos,size)`? We discussed here and in discord, the main consense it, we "expect" from a box normally a pos,size and this is the most common use, and most "wish" the pos,size approach.

But on the other hand, in UE4 and Unity, it's they have the Box2(min,max) approach.

So, it's not a simple choice. I wish a responsible maintainer should give here a strong signal.

7Bpencil commented:
Why opentk has classes like boxes or bezier curve? Vectors and matrixes give nice overrides for GL functions, but these are too high-level. Why not extract them into another library?

NogginBops commented:
@7Bpencil Good question. I think it's all about simplicity.
A lot of people that are going to use OpenTK will be using the provided math structures like `Vector3`.
Then instead of making people find new math packages (or write their own) where stuff like boxes are included we can just include implementations for these structures. Making the life of developers easier.
Boxes and Bezier curves are not that high-level math concepts either, though we are seeing a duality of box representation here where a decision has to be made, but it doesn't really matter that much as the conversion between the two is trivial.

The work to make these structures has already been done so keeping them in the library requires no real maintenance effort.
Also in OpenTK 4.0 the math part of OpenTK is it's own assembly.

(this comment might not be the most thought out answer but it's what I got atm)

NogginBops commented:
Also @varon, @frederikja163, @jvbsl, @PJB3005, one of you should take a stance on the whole `Box` constructor thing that @arakis mentioned above.

jvbsl commented:
Well I don't really know either, but I prefer the min,max one, as everything I've ever worked with(or implemented myself) has done it that way. But that's pretty subjective, an objective view on that would be that, probably most operations on the boxes need less calculations with the min,max approach as well. I Think more opinions would help tbh

7Bpencil commented:
@NogginBops does opentk have plans to give users polygons, spheres, cylinders, etc. etc..?

arakis commented:
@jvbsl i think you're right, and that's what i saw everywhere, too (unread engine 4 and unity use that, too). So, i will continue my PR with the min,max convention as default.

varon commented:
Bump. is this still active? It might be a good candidate for OpenTK 5.0

arakis commented:
Currently, I'm busy for the next three months, if somebody could adapt the PR, it would be nice.

arakis commented:
So, this is a final review-request, befor i will duplicate all the suggested changes to the other box structures.

If you have any complains/improvements, please let me know within the next days. If there are any issues that are not solvable in an accetable way, please suggest to remove/revert them.

NogginBops commented:
@arakis Sorry for not responding to your review-request. I'm not sure why this PR got ignored for such a long time, so sorry about that.
I took at look at the code and I'm happy with the API, so if you still have the time and energy to finish this PR that would be amazing.

arakis commented:
Yeah, i discontinued the work because of lack of feedback :-) But for now, i rebased with upstream. Damm...now the code comments here in the conversation are shifted/gone :-/

arakis commented:
I've finished now the PR. It's just a copy from the Box2.cs to all other structures.

- I did't copy Round, Floor & Ceiling to `Box2i` and `Box3i`.
- I renamed `Zero` to `Empty`, because in the meantime, there was added `Empty` to `Box2i` and `Box3i` by @NogginBops in commit aa3ee95. (Btw: It was only added in theese classes, not to the other ones). So, i renamed `IsZero` to `IsEmtpy`, also.

arakis commented:
I renamed `IsEmpty` back to `IsZero`, like it was before. I'll keep the `Empty` property itself, because it's already there. You can rename it after the Merge (or i rename them....but than we have a breaking change for poeple, they use the `Empty` property since you introduced them).
